Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -1601,36 +1714,92 @@ namespace Tensile
         return size;
     }
 
-    size_t ContractionSolution::getSKGrid(Hardware const& hardware, size_t tiles) const
+    size_t ContractionSolution::getSKGrid(Problem const& problem, Hardware const& hardware, size_t tiles) const
     {
         AMDGPU const* pAMDGPU = dynamic_cast<AMDGPU const*>(&hardware);
-        if(pAMDGPU->skFixedGrid > 0)
-            return pAMDGPU->skFixedGrid;
+        
         assert(pAMDGPU != nullptr && pAMDGPU->computeUnitCount != 0);
         size_t cuCount = pAMDGPU->computeUnitCount;
-        size_t skGrid  = cuCount;
-        if(pAMDGPU->skDynamicGrid == 2 && tiles > skGrid)
+        
+        // User-specified grid size for Stream-K kernel.
+        if(pAMDGPU->skFixedGrid > 0) 
+        {
+            return pAMDGPU->skFixedGrid;
+        }
+
+        // Dynamically pick the minimum between the cuCount or number of tiles.
+        else if(pAMDGPU->skDynamicGrid == 1)
+        {
+            return min(cuCount, tiles);
+        }
+
+        // Dynamically pick the minimum between the cuCount or number of tiles,
+        // and scale down really large sizes to use fewer CUs for power/energy savings.
+        else if(pAMDGPU->skDynamicGrid == 2)
         {
-            for(size_t i = 1; i <= 32; i *= 2)
+            size_t skGrid  = cuCount;
+            if (tiles > skGrid)
             {
-                size_t tilesPerCU  = CeilDivide(i * tiles, skGrid);
-                size_t reducedGrid = CeilDivide(i * tiles, tilesPerCU);
-                float  utilization = ((float)reducedGrid) / ((float)skGrid);
-                if(utilization > 0.75f)
+                for(size_t i = 1; i <= 32; i *= 2)
                 {
-                    if(utilization < 1.0f)
-                        skGrid = reducedGrid;
-                    break;
+                    size_t tilesPerCU  = CeilDivide(i * tiles, cuCount);
+                    size_t reducedGrid = CeilDivide(i * tiles, tilesPerCU);
+                    float  utilization = ((float)reducedGrid) / ((float)cuCount);
+                    if(utilization > 0.75f)
+                    {
+                        if(utilization < 1.0f)
+                            skGrid = reducedGrid;
+                        break;
+                    }
                 }
             }
+
+            return min(skGrid, tiles);
+        }
+
+        // Dynamically predict the best grid-size by weighing the cost of the fix-up
+        // step and the cost of processing MAC-loop instructions. When the cost of fix-up
+        // is the bottleneck, use smaller grid size.
+        // Architecture dependent.
+        else if(pAMDGPU->skDynamicGrid == 3)
+        {
+            dim3 dSize(1, 1, 1);
```

Comment:
I did not clearly mention, but we should not use dim3 here.
The type of each member is uint32.
Please define size_t variables instead.